src/lib.rs

```diff
@@ -158,6 +158,24 @@ impl SpinSleeper {
         }
     }
 
+    /// Puts the [current thread to sleep](fn.native_sleep.html) until instant less
+    /// the configured native accuracy. Then spins until the specified instant is reached.
+    pub fn sleep_until(self, instant: Instant) {
+        let accuracy = Duration::new(0, self.native_accuracy_ns);
+        let start = Instant::now();
+        let duration = instant.duration_since(start);
+        if duration > accuracy {
+            native_sleep(duration - accuracy);
+        }
+        // spin until instant
+        while Instant::now() < instant {
+            match self.spin_strategy {
+                SpinStrategy::YieldThread => thread::yield_now(),
+                SpinStrategy::SpinLoopHint => std::hint::spin_loop(),
+            }
+        }
```

Comment:
Could be implemented simply as:
```suggestion
        self.sleep(instant.saturating_duration_since(Instant::now()))
```

That would mean calling `Instant::now()` twice, but better than what you have since you call `Instant::now()` in a loop.

We can also avoid the 2 calls by having a private fn sleep that takes both duration & start instant.

Reply:
`self.sleep` calls Instant::now() as well, so there would still be two calls. I'm not entirely sure what you mean by the private sleep fn. Wouldn't that just be the same thing?

Reply:
The private fn would exist to solve that problem:
```rust
    fn sleep_from(self, start: Instant, duration: Duration) {
        let accuracy = Duration::new(0, self.native_accuracy_ns);
        if duration > accuracy {
            native_sleep(duration - accuracy);
        }
        // spin the rest of the duration
        while start.elapsed() < duration {
            match self.spin_strategy {
                SpinStrategy::YieldThread => thread::yield_now(),
                SpinStrategy::SpinLoopHint => std::hint::spin_loop(),
            }
        }
    }

    pub fn sleep(self, duration: Duration) {
        self.sleep_from(Instant::now(), duration)
    }

    pub fn sleep_until(self, deadline: Instant) {
        let start = Instant::now();
        self.sleep_from(start, deadline.saturating_duration_since(start))
    }
```

Reply:
`elapsed()` also calls Instant::now().
```rust
pub fn elapsed(&self) -> Duration {
    Instant::now() - *self
}
```

Reply:
At least by comparing `Instant::now()` against `instant` you save an instruction.

```rust
(Instant::now() - *self) < duration
```
compared to
```rust
Instant::now() < instant
```

Reply:
> elapsed() also calls Instant::now().

Yeah but there is no getting around calling now() in the spin loop. I'd rather both methods had common implementations.

Reply:
What if `sleep_until` was the common implementation that `sleep` called?

Reply:
That's possible, but I don't think it makes quite as much sense. The first thing you have to do in `sleep_until` is compute the sleep duration, so we can compare to native sleep accuracy, so "sleeping a duration" seems to be the more founding operation to base both fns on.

`start.elapsed() < duration` vs `Instant::now() < deadline` is an implementation detail that applies to either impl (ie we could change that in `sleep` now regardless of `sleep_until` existence) so I don't think it is relevant.

Reply:
Really my only worry is accuracy. `Instant::now() < deadline` may be more accurate than `start.elapsed() < duration` due to it being one less operation. Maybe there's a way to get the both of both worlds?

Reply:
Sorry, I should have clarified what I mean by accurate. I mean in terms of when the function returns, not accuracy of time measurement. The function would have to do one extra operation before returning for sleep compared to sleep_until.

Reply:
I like the `now < deadline` change, hence https://github.com/alexheretic/spin-sleep/commit/304096f390012e7a909c0f9b682381823e670f1f though I wouldn't be surprised if there is no real compiled difference.

I don't think it is relevant to the code sharing issue that motivated this thread.

Reply:
Would it be such a bad thing to not use shared code? If `sleep_until` relies on `sleep`, there may be an accuracy error from the overhead of calling the function. It would be really hard to test, but it makes sense to me to try to prevent accuracy errors as much as possible. The time between the duration being created in `sleep_until` and `start` being initialized in `sleep` is indeterminate. It's a good strategy for `sleep` itself, but not such a good strategy for `sleep_until`. But it's your repo, so it's really up to you. You could share the spin part between functions since they both rely on a deadline.  So long as the function that spins is marked as inline, it should be an optimized solution for both cases.

Reply:
I'm not sure how to link commits like what you did, but [this](https://github.com/alexheretic/spin-sleep/pull/28/commits/b39bd5fad382e3b9b605b1f3d21be9633d40d788) commit shows the `spin_sleep` functionality, which can be shared between both methods since they both create a duration and a deadline.